// astro/dist/core/build/page-data.js
import { fileURLToPath } from "url";
import * as colors from "kleur/colors";
import { debug } from "../logger.js";
import { preload as ssrPreload } from "../ssr/index.js";
import { validateGetStaticPathsModule, validateGetStaticPathsResult } from "../ssr/routing.js";
import { generateRssFunction } from "../ssr/rss.js";
import { assignStaticPaths } from "../ssr/route-cache.js";
async function collectPagesData(opts) {
  const { astroConfig, logging, manifest, origin, routeCache, viteServer } = opts;
  const assets = {};
  const allPages = {};
  await Promise.all(manifest.routes.map(async (route) => {
    var _a, _b;
    if (route.pathname) {
      allPages[route.component] = {
        route,
        paths: [route.pathname],
        preload: await ssrPreload({
          astroConfig,
          filePath: new URL(`./${route.component}`, astroConfig.projectRoot),
          logging,
          mode: "production",
          origin,
          pathname: route.pathname,
          route,
          routeCache,
          viteServer
        }).then((routes) => {
          const html = `${route.pathname}`.replace(/\/?$/, "/index.html");
          debug(logging, "build", `\u251C\u2500\u2500 ${colors.bold(colors.green("\u2714"))} ${route.component} \u2192 ${colors.yellow(html)}`);
          return routes;
        }).catch((err) => {
          debug(logging, "build", `\u251C\u2500\u2500 ${colors.bold(colors.red("\u2718"))} ${route.component}`);
          throw err;
        })
      };
      return;
    }
    const result = await getStaticPathsForRoute(opts, route).then((routes) => {
      const label = routes.paths.length === 1 ? "page" : "pages";
      debug(logging, "build", `\u251C\u2500\u2500 ${colors.bold(colors.green("\u2714"))} ${route.component} \u2192 ${colors.magenta(`[${routes.paths.length} ${label}]`)}`);
      return routes;
    }).catch((err) => {
      debug(logging, "build", `\u251C\u2500\u2500 ${colors.bold(colors.red("\u2717"))} ${route.component}`);
      throw err;
    });
    if ((_a = result.rss) == null ? void 0 : _a.length) {
      for (let i = 0; i < result.rss.length; i++) {
        const rss = result.rss[i];
        if (rss.xml) {
          const { url, content } = rss.xml;
          if (content) {
            const rssFile = new URL(url.replace(/^\/?/, "./"), astroConfig.dist);
            if (assets[fileURLToPath(rssFile)]) {
              throw new Error(`[getStaticPaths] RSS feed ${url} already exists.
Use \`rss(data, {url: '...'})\` to choose a unique, custom URL. (${route.component})`);
            }
            assets[fileURLToPath(rssFile)] = content;
          }
        }
        if ((_b = rss.xsl) == null ? void 0 : _b.content) {
          const { url, content } = rss.xsl;
          const stylesheetFile = new URL(url.replace(/^\/?/, "./"), astroConfig.dist);
          if (assets[fileURLToPath(stylesheetFile)]) {
            throw new Error(`[getStaticPaths] RSS feed stylesheet ${url} already exists.
Use \`rss(data, {stylesheet: '...'})\` to choose a unique, custom URL. (${route.component})`);
          }
          assets[fileURLToPath(stylesheetFile)] = content;
        }
      }
    }
    allPages[route.component] = {
      route,
      paths: result.paths,
      preload: await ssrPreload({
        astroConfig,
        filePath: new URL(`./${route.component}`, astroConfig.projectRoot),
        logging,
        mode: "production",
        origin,
        pathname: result.paths[0],
        route,
        routeCache,
        viteServer
      })
    };
  }));
  return { assets, allPages };
}
async function getStaticPathsForRoute(opts, route) {
  const { astroConfig, logging, routeCache, viteServer } = opts;
  if (!viteServer)
    throw new Error(`vite.createServer() not called!`);
  const filePath = new URL(`./${route.component}`, astroConfig.projectRoot);
  const mod = await viteServer.ssrLoadModule(fileURLToPath(filePath));
  validateGetStaticPathsModule(mod);
  const rss = generateRssFunction(astroConfig.buildOptions.site, route);
  await assignStaticPaths(routeCache, route, mod, rss.generator);
  const staticPaths = routeCache[route.component];
  validateGetStaticPathsResult(staticPaths, logging);
  return {
    paths: staticPaths.map((staticPath) => staticPath.params && route.generate(staticPath.params)).filter(Boolean),
    rss: rss.rss
  };
}
export {
  collectPagesData
};
