// astro/dist/core/build/vite-plugin-hoisted-scripts.js
function virtualHoistedEntry(id) {
  return id.endsWith(".astro/hoisted.js") || id.endsWith(".md/hoisted.js");
}
function vitePluginHoistedScripts(internals) {
  return {
    name: "@astro/rollup-plugin-astro-hoisted-scripts",
    resolveId(id) {
      if (virtualHoistedEntry(id)) {
        return id;
      }
    },
    load(id) {
      if (virtualHoistedEntry(id)) {
        let code = "";
        for (let path of internals.hoistedScriptIdToHoistedMap.get(id)) {
          code += `import "${path}";`;
        }
        return {
          code
        };
      }
      return void 0;
    },
    async generateBundle(_options, bundle) {
      for (const [id, output] of Object.entries(bundle)) {
        if (output.type === "chunk" && output.facadeModuleId && virtualHoistedEntry(output.facadeModuleId)) {
          const facadeId = output.facadeModuleId;
          const filename = facadeId.slice(0, facadeId.length - "/hoisted.js".length);
          internals.facadeIdToHoistedEntryMap.set(filename, id);
        }
      }
    }
  };
}
export {
  vitePluginHoistedScripts
};
