// astro/dist/core/ssr/route-cache.js
import { debug } from "../logger.js";
import { generatePaginateFunction } from "../ssr/paginate.js";
async function callGetStaticPaths(mod, route, rssFn) {
  const staticPaths = await (await mod.getStaticPaths({
    paginate: generatePaginateFunction(route),
    rss: rssFn || (() => {
    })
  })).flat();
  const keyedStaticPaths = staticPaths;
  keyedStaticPaths.keyed = new Map();
  for (const sp of keyedStaticPaths) {
    const paramsKey = JSON.stringify(sp.params);
    keyedStaticPaths.keyed.set(paramsKey, sp);
  }
  return keyedStaticPaths;
}
async function assignStaticPaths(routeCache, route, mod, rssFn) {
  const staticPaths = await callGetStaticPaths(mod, route, rssFn);
  routeCache[route.component] = staticPaths;
}
async function ensureRouteCached(routeCache, route, mod, rssFn) {
  if (!routeCache[route.component]) {
    const staticPaths = await callGetStaticPaths(mod, route, rssFn);
    routeCache[route.component] = staticPaths;
    return staticPaths;
  } else {
    return routeCache[route.component];
  }
}
function findPathItemByKey(staticPaths, paramsKey, logging) {
  let matchedStaticPath = staticPaths.keyed.get(paramsKey);
  if (matchedStaticPath) {
    return matchedStaticPath;
  }
  debug(logging, "findPathItemByKey", `Unexpected cache miss looking for ${paramsKey}`);
  matchedStaticPath = staticPaths.find(({ params: _params }) => JSON.stringify(_params) === paramsKey);
}
export {
  assignStaticPaths,
  callGetStaticPaths,
  ensureRouteCached,
  findPathItemByKey
};
