// astro/dist/core/preview/index.js
import http from "http";
import { performance } from "perf_hooks";
import send from "send";
import { fileURLToPath } from "url";
import fs from "fs";
import * as msg from "../dev/messages.js";
import { error, info } from "../logger.js";
import { subpathNotUsedTemplate, notFoundTemplate } from "../dev/template/4xx.js";
import { appendForwardSlash, trimSlashes } from "../path.js";
async function preview(config, { logging }) {
  const startServerTime = performance.now();
  const pageUrlFormat = config.buildOptions.pageUrlFormat;
  const trailingSlash = config.devOptions.trailingSlash;
  const forceTrailingSlash = trailingSlash === "always";
  const blockTrailingSlash = trailingSlash === "never";
  const defaultFile = "index.html";
  const defaultOrigin = "http://localhost";
  const sendOptions = {
    extensions: pageUrlFormat === "file" ? ["html"] : false,
    index: false,
    root: fileURLToPath(config.dist)
  };
  let baseURL = new URL(appendForwardSlash(config.buildOptions.site || ""), defaultOrigin);
  const server = http.createServer((req, res) => {
    const requestURL = new URL(req.url, defaultOrigin);
    if (!requestURL.pathname.startsWith(baseURL.pathname)) {
      res.statusCode = 404;
      res.end(subpathNotUsedTemplate(baseURL.pathname, requestURL.pathname));
      return;
    }
    const pathname = requestURL.pathname.slice(baseURL.pathname.length - 1);
    const isRoot = pathname === "/";
    const hasTrailingSlash = isRoot || pathname.endsWith("/");
    let tryTrailingSlash = true;
    let tryHtmlExtension = true;
    let url;
    const onErr = (message) => {
      res.statusCode = 404;
      res.end(notFoundTemplate(pathname, message));
    };
    const onStat = (err, stat) => {
      switch (true) {
        case (err && tryHtmlExtension && hasTrailingSlash && !blockTrailingSlash):
        case (err && tryHtmlExtension && !hasTrailingSlash && !forceTrailingSlash && !pathname.endsWith(".html")):
          tryHtmlExtension = false;
          return fs.stat(url = new URL(url.pathname + ".html", url), onStat);
        case err !== null:
          return onErr("Path not found");
        case (stat.isDirectory() && hasTrailingSlash && blockTrailingSlash && !isRoot):
          return onErr("Prohibited trailing slash");
        case (stat.isDirectory() && !hasTrailingSlash && forceTrailingSlash && !isRoot):
          return onErr("Required trailing slash");
        case (stat.isDirectory() && tryTrailingSlash):
          tryTrailingSlash = false;
          return fs.stat(url = new URL(url.pathname + (url.pathname.endsWith("/") ? defaultFile : "/" + defaultFile), url), onStat);
        case stat.isDirectory():
          return onErr("Path not found");
        default:
          send(req, fileURLToPath(url), {
            extensions: false,
            index: false
          }).pipe(res);
      }
    };
    fs.stat(url = new URL(trimSlashes(pathname), config.dist), onStat);
  });
  let { hostname, port } = config.devOptions;
  let httpServer;
  function startServer(timerStart) {
    let showedPortTakenMsg = false;
    let showedListenMsg = false;
    return new Promise((resolve, reject) => {
      const listen = () => {
        httpServer = server.listen(port, hostname, () => {
          if (!showedListenMsg) {
            info(logging, "astro", msg.devStart({ startupTime: performance.now() - timerStart }));
            info(logging, "astro", msg.devHost({ host: `http://${hostname}:${port}${baseURL.pathname}` }));
          }
          showedListenMsg = true;
          resolve();
        });
        httpServer == null ? void 0 : httpServer.on("error", onError);
      };
      const onError = (err) => {
        if (err.code && err.code === "EADDRINUSE") {
          if (!showedPortTakenMsg) {
            info(logging, "astro", msg.portInUse({ port }));
            showedPortTakenMsg = true;
          }
          port++;
          return listen();
        } else {
          error(logging, "astro", err.stack);
          httpServer == null ? void 0 : httpServer.removeListener("error", onError);
          reject(err);
        }
      };
      listen();
    });
  }
  await startServer(startServerTime);
  return {
    hostname,
    port,
    server: httpServer,
    stop: async () => {
      httpServer.close();
    }
  };
}
export {
  preview as default
};
