// astro/dist/vite-plugin-astro/index.js
import esbuild from "esbuild";
import npath from "path";
import { fileURLToPath } from "url";
import { getViteTransform } from "./styles.js";
import { parseAstroRequest } from "./query.js";
import { cachedCompilation, invalidateCompilation } from "./compile.js";
import ancestor from "common-ancestor-path";
const FRONTMATTER_PARSE_REGEXP = /^\-\-\-(.*)^\-\-\-/ms;
function astro({ config, logging }) {
  function normalizeFilename(filename) {
    if (filename.startsWith("/@fs")) {
      filename = filename.slice("/@fs".length);
    } else if (filename.startsWith("/") && !ancestor(filename, config.projectRoot.pathname)) {
      filename = new URL("." + filename, config.projectRoot).pathname;
    }
    return filename;
  }
  let viteTransform;
  const srcRootWeb = config.src.pathname.slice(config.projectRoot.pathname.length - 1);
  const isBrowserPath = (path) => path.startsWith(srcRootWeb);
  return {
    name: "@astrojs/vite-plugin-astro",
    enforce: "pre",
    configResolved(resolvedConfig) {
      viteTransform = getViteTransform(resolvedConfig);
    },
    async resolveId(id) {
      const { query } = parseAstroRequest(id);
      if (query.astro) {
        if (query.type === "style" && isBrowserPath(id)) {
          const outId = npath.posix.join(config.projectRoot.pathname, id);
          return outId;
        }
        return id;
      }
    },
    async load(id, opts) {
      let { filename, query } = parseAstroRequest(id);
      if (query.astro) {
        if (query.type === "style") {
          if (typeof query.index === "undefined") {
            throw new Error(`Requests for Astro CSS must include an index.`);
          }
          const transformResult = await cachedCompilation(config, normalizeFilename(filename), null, viteTransform, opts);
          const csses = transformResult.css;
          const code = csses[query.index];
          return {
            code
          };
        } else if (query.type === "script") {
          if (typeof query.index === "undefined") {
            throw new Error(`Requests for hoisted scripts must include an index`);
          }
          const transformResult = await cachedCompilation(config, normalizeFilename(filename), null, viteTransform, opts);
          const scripts = transformResult.scripts;
          const hoistedScript = scripts[query.index];
          if (!hoistedScript) {
            throw new Error(`No hoisted script at index ${query.index}`);
          }
          return {
            code: hoistedScript.type === "inline" ? hoistedScript.code : `import "${hoistedScript.src}";`
          };
        }
      }
      return null;
    },
    async transform(source, id, opts) {
      if (!id.endsWith(".astro")) {
        return;
      }
      try {
        const transformResult = await cachedCompilation(config, id, source, viteTransform, opts);
        const { code, map } = await esbuild.transform(transformResult.code, {
          loader: "ts",
          sourcemap: "external",
          sourcefile: id
        });
        return {
          code,
          map
        };
      } catch (err) {
        const scannedFrontmatter = FRONTMATTER_PARSE_REGEXP.exec(source);
        if (scannedFrontmatter) {
          try {
            await esbuild.transform(scannedFrontmatter[1], { loader: "ts", sourcemap: false, sourcefile: id });
          } catch (frontmatterErr) {
            if (frontmatterErr && frontmatterErr.message) {
              frontmatterErr.message = frontmatterErr.message.replace("end of file", "end of frontmatter");
            }
            throw frontmatterErr;
          }
        }
        if (err.stack.includes("wasm-function")) {
          const search = new URLSearchParams({
            labels: "compiler",
            title: "\u{1F41B} BUG: `@astrojs/compiler` panic",
            body: `### Describe the Bug
    
    \`@astrojs/compiler\` encountered an unrecoverable error when compiling the following file.
    
    **${id.replace(fileURLToPath(config.projectRoot), "")}**
    \`\`\`astro
    ${source}
    \`\`\`
    `
          });
          err.url = `https://github.com/withastro/astro/issues/new?${search.toString()}`;
          err.message = `Error: Uh oh, the Astro compiler encountered an unrecoverable error!
    
    Please open
    a GitHub issue using the link below:
    ${err.url}`;
          if (logging.level !== "debug") {
            err.stack = `    at ${id}`;
          }
        }
        throw err;
      }
    },
    async handleHotUpdate(context) {
      invalidateCompilation(config, context.file);
    }
  };
}
export {
  astro as default
};
